Kumi.Game/IO/Formats/FileDecoder.cs

```diff
@@ -0,0 +1,294 @@
+    }
+    
+    private string sanitizeHeader(string line)
+    {
+        if (!isSectionHeader(line))
+            return line;
+
+        var sectionName = stripHeader(line);
+        
+        if (sectionName.IsPascalCase())
+            return line;
+
+        if (sectionName.ToLower().IsSnakeCase())
+        {
+            var sb = new StringBuilder();
+            var words = sectionName.ToLower().Split('_');
+            foreach (var word in words)
+            {
+                sb.Append(char.ToUpper(word[0]));
+                sb.Append(word.Substring(1));
+            }
+
+            return $"{sb}";
+        }
+
+        if (sectionName.IsCamelCase())
+        {
+            // Should just be able to capitalize the first letter.
+            return $"{char.ToUpper(sectionName[0])}{sectionName.Substring(1)}";
+        }
+        
+        // If it's not camel or snake case, just return the line as-is.
+        // I'm sure we're past the point of caring anymore.
+        return sectionName;
+    }
+    
+    private string stripHeader(string line)
+    {
+        if (!isSectionHeader(line))
+            return line;
+
+        return line.Substring(SectionHeader.Start.Length, line.Length - SectionHeader.End.Length - SectionHeader.Start.Length);
+    }
+
+    /// <summary>
+    /// A class containing the start and end characters of a section header.
+    /// </summary>
+    protected class SectionHeaderValues
+    {
+        public string Start { get; set;  }
+        public string End { get; set;  }
+    }
+}
+
+/// <summary>
+/// A line-based file decoder for a specific file format.
+/// </summary>
+/// <typeparam name="T"></typeparam>
+public class FileDecoder<T> : IFileHandler<T>
+    where T : new()
+{
+    /// <summary>
+    /// The current version of the file format.
+    /// </summary>
+    public int Version { get; }
+    
+    public T Current { get; set; } = new T();
+
+    /// <summary>
+    /// The characters used to denote a comment.
+    /// </summary>
+    protected string CommentCharacter { get; } = "#";
+    
+    protected FileDecoder(int version)
+    {
+        Version = version;
+    }
+    
+    public T Decode(Stream stream) => Decode(stream, new T());
+
+    /// <summary>
+    /// Parses a readable stream into a designated output of type <see cref="T"/>.
+    /// </summary>
+    /// <param name="stream">The stream to parse.</param>
+    public virtual T Decode(Stream stream, T input)
+    {
+        using var reader = new LineBufferedReader(stream);
+        
+        string header = reader.ReadLine();
+        if (!ValidateHeader(header))
+            throw new InvalidDataException($"The header of the file being parsed by {nameof(FileDecoder<T>)} is invalid: {header}");
+        
+        Current = input;
+        string line;
+        
+        while ((line = reader.ReadLine()) != null)
+        {
+            if (ShouldIgnoreLine(line))
+                continue;
+
+            line = line.TrimEnd();
```

Comment:
```suggestion
            line = line.Trim();
```

Reply:
don't know how the end got added because it was trim originally